### PR TITLE
[persist] Use the new MFP filtering in the persist source

### PR DESCRIPTION
**Pull request by bkirwi**
Following up on #18932, wire the new interpreter into the source.

### Motivation

Known-desireable: #18007.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:157]: nit: feels like this has passed the complexity threshold for being inlined in the closure definition. can we move the whole block out into a freestanding function?

bonus super nit: let's early return on the `if let Some(plan) = &filter_plan` by turning it in to a match with one case returning

(I think we actually want to move the bulk of this down into mz_expr somewhere, but that should be left for a later PR)

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:195]: this per-column block of ResultSpec logic reads beautifully :clap:

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:398]: i have a preference that this instead match on the `ColumnType`, which would let us be loud about unexpected stats types in the future, but feel free to ignore this for this PR if you like, and we can circle back later

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:159]: I think this all needs to be fake_desc instead of desc

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:441]: can `JsonStats::None` be `ResultSpec::nothing`?

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:434]: is this still true? let's move it to a TODO(mfp) if it is and if we think there's some way to avoid this with more info in the stats

Comment (bkirwi) [reply src/storage-client/src/source/persist_source.rs:434]: Still true I think! Changed the comment.

File: src/storage-client/src/source/persist_source.rs
Comment (bkirwi) [reply src/storage-client/src/source/persist_source.rs:441]: I wasn't sure if we wanted to reserve the right to use`JsonStats::None` when we don't have stats for some reason. (vs. for eg. an empty JSON column.) In which case, should be fine?

File: src/storage-client/src/source/persist_source.rs
Comment (bkirwi) [reply src/storage-client/src/source/persist_source.rs:159]: Ugh, this was intentional but I can no longer remember why! If I can't figure it out I'll switch it back.

File: src/storage-client/src/source/persist_source.rs
Comment (bkirwi) [reply src/storage-client/src/source/persist_source.rs:398]: What are the odds that we'd ever track JSON stats for non-bytes columns? If small I'd prefer to avoid having to maintain the exhaustive match. (If non-small, agreed!)

File: src/storage-client/src/source/persist_source.rs
Comment (bkirwi) [reply src/storage-client/src/source/persist_source.rs:159]: On reflection I think the rationale is that by the time we're calling `MFP::evaluate` on a particular `Row`, the type of the row is described by the actual `RelationType` and not the hack. So I think this is the correct thing!

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:163]: it'd be nice if this could take a ref to desc instead of cloning

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [reply src/storage-client/src/source/persist_source.rs:408]: hmm, not sure if I got my point across :). it feels like an error to me (and we probably should make it at least a debug_assert) for a caller of `StructStats::col` to request stats for a column using a type other than that column's type. for example (but without loss on generality), the error path for that method has a `format!` in it, which is reasonable to run in a true error case, but not reasonable to run in the common case

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [reply src/storage-client/src/source/persist_source.rs:451]: not entirely sure I follow :)

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:444]: ah, I see now! we probably need an equivalent to StructStats::len on JsonStats::Maps

(maybe this also clarifies my JsonStats::None question, too? i.e. the len field on StructStats is why we don't need a sentinel for what we've pruned out of the StructStats)

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:173]: I mentioned in my last round of comments that we might want to later move this method into `mz_expr` at some point... but I just realized that maybe you are ahead of me here! :D

I intended for the `PersistSourceDataStats` trait to be the API boundary between persist and compute, but looking at this again, I had a 💡 go on and realized that the interpreter stuff is (as of this PR) effectively the API boundary instead. which might be a nicer API boundary! is that what you've been thinking?

**Comment by bkirwi** [comment]
@danhhz and I discussed his comments elsewhere; I've pushed some changes up in the last three commits.

I'm re-marking this as ready for review, now that initial feedback is addressed and we have a fix for the interpreter bug: https://github.com/MaterializeInc/materialize/pull/19173.

**Review by danhhz — APPROVED** [review]

**Comment by bkirwi** [comment]
Recording an observation from @danhhz earlier today:
- The bug fixed in #19173 involves the interpreter will sometimes return a fallible spec for an expression that will never actually fail.
- If the part filter receives a fallible spec, it always keeps the part.

So even though the interpreter is incorrect, it fails in a way that does not affect the correctness of the part filter.

I will probably still wait until #19173 is in before merging this. But if it looks like #19173 may not hit the release cutoff, we may merge this one first to unblock some additional testing.

**Pull request merged**
